pax_achievements.py

```diff
@@ -1,434 +1,258 @@
+
+
+def posts(df: pd.DataFrame, bb_filter: pd.Series, ao_filter: pd.Series) -> pd.DataFrame:
+    grouping = ["year", "slack_user_id", "home_region"]
+    x = (
+        df.assign(year=df.date.dt.year)
+        .query("not (@bb_filter or @ao_filter)")
+        .groupby(grouping)
+        .agg({"ao_id": "count", "date": "max"})
+    )
+    return x
+
+
+def six_pack(df: pd.DataFrame, bb_filter: pd.Series, ao_filter: pd.Series) -> pd.DataFrame:
+    grouping = ["week", "slack_user_id", "home_region"]
+    x = (
+        df.assign(week=df.date.dt.isocalendar().week)
+        .query("not (@bb_filter or @ao_filter)")
+        .groupby(grouping)
+        .agg({"ao_id": "count", "date": "max"})
+    )
+    return x[x.ao_id >= 6].reset_index().drop("ao_id", axis=1).rename({"date": "date_awarded"}, axis=1)
+
+
+def hdtf(df: pd.DataFrame, bb_filter: pd.Series, ao_filter: pd.Series) -> pd.DataFrame:
+    grouping = ["year", "slack_user_id", "home_region", "ao_id"]
+    x = (
+        df.assign(year=df.date.dt.year)
+        .query("not (@bb_filter or @ao_filter)")
+        .groupby(grouping)
+        .agg({"ao": "count", "date": "max"})
+    )
+    return x[x.ao >= 50].reset_index().drop("ao", axis=1).rename({"date": "date_awarded"}, axis=1)
+
+
+def load_to_database(row, engine: Engine, metadata: MetaData, data_to_load: pd.DataFrame) -> None:
+    """after successfully sending Slack notifications, push the data to the `achievements_awarded` table.
+    The data frame data_to_load has already been filtered to include only new achievements."""
+    try:
+        aa = metadata.tables[f"{row.paxminer_schema}.achievements_awarded"]
+    except KeyError:
+        aa = Table("achievements_awarded", metadata, autoload_with=engine, schema=row.paxminer_schema)
+
+    load_records = data_to_load.to_dict("records")
+    sql = insert(aa).values(load_records)
+    with engine.begin() as cnxn:
+        cnxn.execute(sql)
+
+
+def main():
+    year = date.today().year
+    engine = mysql_connection()
+    metadata = MetaData()
+    metadata.reflect(engine, schema="weaselbot")
+
+    df_regions = pd.read_sql(region_sql(metadata), engine)
+    nation_df = pd.read_sql(nation_sql(metadata, year), engine, parse_dates="date")
+
+    # for QSource, we want to capture only QSource
+    bb_filter = nation_df.backblast.str.lower().str[:100].str.contains(r"q.{0,1}source|q{0,1}[1-9]\.[0-9]\s")
+    ao_filter = nation_df.ao.str.contains(r"q.{0,1}source")
+
+    dfs = []
+    ############# Q Source ##############
+    dfs.append(the_priest(nation_df, bb_filter, ao_filter))
+    dfs.append(the_monk(nation_df, bb_filter, ao_filter))
+    ############### END #################
+
+    # For beatdowns, we want to exclude QSource and Ruck (blackops too? What is blackops?)
+    bb_filter = nation_df.backblast.str.lower().str[:100].str.contains(r"q.{0,1}source|q{0,1}[1-9]\.[0-9]\s|ruck")
+    ao_filter = nation_df.ao.str.contains(r"q.{0,1}source|ruck")
```

Comment:
"blackops" aka "off the record" aka probably a million other names is what we use for non-scheduled workouts. When I initially set all this up for my region, we decided that these should not be included as a normal beatdown (we had guys that would ruck a mile and call it a beatdown). Most regions probalby don't even use the term blackops, so this probably only applied to a small amount of regions... we should think about how to handle this going forward

Reply:
For sure. This is nebulous for sure. We call it "OTB" but men here are held accountable to listing these in QSignups, holding them to a minimum of 45 minutes and treat them like any other workout. So we count them. But I can see how discipline might wane. Leaving it for now but will monitor and continue to think it through.